### PR TITLE
Load closed shops in a separate request on /shops page

### PR DESCRIPTION
**Pull request by Matt-Yorkley**
#### What? Why?

Part of #5143

Loads only open shops by default on the `/shops` page, and loads closed shops separately when the "Show Closed Shops" button is clicked.

#### What should we test?

- `/shops` page loads a lot faster than before. 
- Closed shops are displayed as before, after the "Show Closed Shops" button is clicked. 
- Filtering works as before.

#### Release notes

Load closed shops separately on /shops page

Changelog Category: Changed

### CODE REVIEW

**Comment by luisramos0** [comment]
This makes me think about API first: what is the enterprises and shops endpoints we want to have?

Maybe we should separate enterprises from shops? An enterprise includes profiles and producers, etc Shops must have respective OrderCycle? Or are all producers closed_shops?

I think the concept of "recently closed shops" is important. An enterprise that never had an OC is very different from an enterprise that just closed it's weekly OC on a Friday and will open in on Tuesday as usual. The API should be prepared to take this.

Maybe we can have enterprises list all enterprises:
/api/enterprises list all enterprises
and then have shops endpoints:
/api/enterprises/shops  lists all open shops
/api/enterprises/shops/closed  lists all enterprises not ready for checkout

what do you think?

**Comment by Matt-Yorkley** [comment]
> Maybe we should separate enterprises from shops?

Yes, I was thinking about looking at that (after making sure it all works). We now have two "shop" actions in `api/enterprises_controller`; `shopfront` (which is basically a show action) and `closed_shops`... maybe we could move these to a new `api/shops_controller` ? 

Both of these actions require this in the enterprise controller: `skip_authorization_check only: [:shopfront, :closed_shops]` but all the other actions don't. 

I feel like `api/enterprise_controller` is for CRUD actions on the actual enterprise resource (with authorisation), whereas these shop actions are for reading publicly-accessible representations of an enterprise that is selling to customers.

> I think the concept of "recently closed shops" is important. An enterprise that never had an OC is very different from an enterprise that just closed it's weekly OC on a Friday and will open in on Tuesday as usual. The API should be prepared to take this.

In theory yes, but we don't currently have anything that needs that specific info from the API, so I'd say YAGNI...

> Maybe we can have enterprises list all enterprises: ... /api/enterprises list all enterprises

I don't think we have a current use for that endpoint either...

**Comment by luisramos0** [comment]
awesome! yes to api/shops_controller!

we dont have /api/enterprises, the list, just because we do inject_enterprises, otherwise we would need it. It's one of those that we should have, it's a very basic feature missing from the OFN api right now: list all enterprises (regardless of open/closed shops).

**Comment by kirstenalarsen** [comment]
agree with @luisramos0 - having enterprises on the API would be ace - not least opening up our abilities to simply pull that information into other, better directory and mapping interfaces - so that we could potentially quickly and cheaply have a much more effective directory function :)

File: app/controllers/api/shops_controller.rb [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+    def show
+      enterprise = Enterprise.find_by_id(params[:id])
+
+      render text: Api::EnterpriseShopfrontSerializer.new(enterprise).to_json, status: :ok
+    end
+
+    def closed_shops
+      @active_distributor_ids = []
+      @earliest_closing_times = []
```
[the remark this answers is not in the document]
Comment (Matt-Yorkley) [reply app/controllers/api/shops_controller.rb:16]: These are used in `OpenFoodNetwork::EnterpriseInjectionData.new` a few lines later, and will trigger big queries (that we don't need in the case of closed shops) unless these two ivars are already set.

**Comment by Matt-Yorkley** [comment]
@mkllnk 

> ...three buttons show for a very short time

Fixed :+1:

> A loading indicator would be awesome

Added :+1:

> the filter options are not updating correctly

Give it another try. The taxons "Filter by Type" are loading correctly and working. Are they set up correctly in yoyr environment? Clicking the "Filter by Property" buttons also works correctly...

File: app/assets/javascripts/darkswarm/controllers/hub_node_controller.js.coffee [hunk 2]
```diff
@@ -24,7 +24,7 @@ Darkswarm.controller "HubNodeCtrl", ($scope, HashNavigation, CurrentHub, $http,
     $scope.shopfront_loading = true
     $scope.toggle_tab(event)
 
-    $http.get("/api/enterprises/" + $scope.hub.id + "/shopfront")
+    $http.get("/api/shops/" + $scope.hub.id)
```
Comment (sauloperez) [inline app/assets/javascripts/darkswarm/controllers/hub_node_controller.js.coffee:27]: :heart_eyes:

File: app/controllers/api/shops_controller.rb [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+    respond_to :json
+    skip_authorization_check only: [:show, :closed_shops]
+
+    def show
+      enterprise = Enterprise.find_by_id(params[:id])
+
+      render text: Api::EnterpriseShopfrontSerializer.new(enterprise).to_json, status: :ok
+    end
+
+    def closed_shops
```
Comment (sauloperez) [inline app/controllers/api/shops_controller.rb:14]: IMO this should be a filter specified by URL query param for an `API::ShopsController#index` so we follow Rails REST conventions. But I want to discuss this in another PR and without any rush. This is awesome as it is.

**Comment by filipefurtad0** [comment]
Hey @Matt-Yorkley,

I had a look at the upload times of the /shops page, including closed shops, before and after staging this in UK-staging. It was in both cases around 30s, but I noticed my CPUs were over the top, so perhaps the limiting step is my system.

After UK-staging:
- Type, Property and Delivery filters worked fine, showing both open and closed shops. I also had a closer look at the "Type" filter, since @mkllnk observed some issues before. I chose different combinations, found no issues.
- Closed shops can be toggled, as usual.

Good to go.

**Pull request merged**